### PR TITLE
Generate forward C++ Operator code for wrapped PyTorch modules

### PR DESCRIPTION
**Pull request by orausch**

### CODE REVIEW

File: daceml/onnx/onnx_importer.py [hunk 1]
```diff
@@ -423,18 +431,17 @@ def compile_and_init(self) -> compiled_sdfg.CompiledSDFG:
...
         for name, arr in self.weights.items():
-            desc = self.sdfg.arrays[clean_onnx_name(name)]
-            if type(desc) is dt.Scalar:
-                self.initialized_parameters[clean_onnx_name(
-                    name)] = arr.cpu().numpy()[()]
-            else:
-                cuda = is_cuda(desc.storage)
-                self.initialized_parameters[clean_onnx_name(
-                    name)] = arr.cuda() if cuda else arr
+            if clean_onnx_name(name) in self.sdfg.arrays:
```
Comment (tbennun) [inline daceml/onnx/onnx_importer.py:434]: What happens if it isn't?

File: daceml/pytorch/module_codegen.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+
+def constant_initializer_code(name: str, desc: data.Data, value) -> str:
+    if isinstance(desc, data.Array):
+        iterator = np.nditer(value.cpu().numpy(), order="C")
+        return f"{desc.dtype.ctype} {name}[{desc.total_size}] = {{{', '.join(str(e) for e in iterator)}}};"
+    else:
+        return f"{desc.dtype.ctype} {name} = {str(value.item())};"
+
+
+def code_for_backward_function(module: 'daceml.pytorch.DaceModule',
```
Comment (tbennun) [inline daceml/pytorch/module_codegen.py:141]: do you plan to generate both in the same file?

File: daceml/pytorch/module_codegen.py [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+TORCH_LIBRARY(daceml_{sdfg_name}, m) {{
+    m.def("{sdfg_name}(int handle_ptr, {", ".join('Tensor ' + arg for arg in inputs)}) -> {'Tensor' if len(outputs) == 1
+        else "(" + ", ".join(['Tensor'] * len(outputs)) + ")"}");
+}}
+
+// function definition
+{"Tensor" if len(outputs) == 1 else f"std::tuple<{', '.join(['Tensor'] * len(outputs))}>"}
+{sdfg_name}(int64_t handle_ptr, {",".join(f"const Tensor& {name}_" for name in inputs)}) {{
+
+    // TODO check contiguous
```
Comment (tbennun) [inline daceml/pytorch/module_codegen.py:180]: important, but can be split into two: inputs can be checked for `is_contiguous()` in C++, and parameters are done on the python side prior, to get strides?

File: daceml/pytorch/module_codegen.py [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+        else f"{{{', '.join(o + '_' for o in outputs)}}}"};
+}}
+
+TORCH_LIBRARY_IMPL(daceml_{sdfg_name}, {'CUDA' if module.cuda else 'CPU'}, m) {{
+    m.impl("{sdfg_name}", {sdfg_name});
+}}
+        """
+
+
+def get_function_for_module(module: 'daceml.pytorch.DaceModule',
```
Comment (tbennun) [inline daceml/pytorch/module_codegen.py:204]: I'd rename this to `compile_and_get_function` to make it clear that this compiles the SDFG.

File: daceml/transformation/constant_folding.py [hunk 5]
```diff
@@ -218,3 +218,13 @@ def remove_node_and_computation(sdfg: dace.SDFG, state: dace.SDFGState,
...
                     next_node)) == 0 and not data_used_in_other_states:
                 queue.append(next_node)
+
+    # remove all now useless data descriptors
+    all_read_or_written_data = set(e.data.data for s in sdfg.nodes()
+                                   for e in s.edges())
+    all_read_or_written_data = all_read_or_written_data.union(
+        node.data for node, _ in sdfg.all_nodes_recursive()
+        if isinstance(node, nd.AccessNode))
+    to_delete = set(sdfg.arrays).difference(all_read_or_written_data)
```
Comment (tbennun) [inline daceml/transformation/constant_folding.py:228]: I use `set(sdfg.arrays) - all_read_or_written_data` for clarity

File: tests/pytorch/test_conv2d.py [hunk 6]
```diff
@@ -1,5 +1,6 @@
 import torch
 import torch.nn as nn
+import faulthandler
```
Comment (tbennun) [inline tests/pytorch/test_conv2d.py:3]: ?

**Review by tbennun — CHANGES_REQUESTED** [review]
Some comments that need answering and minor modifications.

File: daceml/pytorch/module_codegen.py (hunk 3 above)
Comment (orausch) [reply daceml/pytorch/module_codegen.py:180]: good point

File: daceml/onnx/onnx_importer.py (hunk 1 above)
Comment (orausch) [reply daceml/onnx/onnx_importer.py:434]: All weights are added to the SDFG on import. If it's missing here, it means a transformation removed it

File: daceml/pytorch/module_codegen.py (hunk 2 above)
Comment (orausch) [reply daceml/pytorch/module_codegen.py:159]: depends on how much code is shared and how long is. At the moment I think so

**Review by tbennun — APPROVED** [review]

**Pull request merged**